modules/extensions-soap-support/src/main/java/org/mule/runtime/module/extension/soap/internal/runtime/connection/ForwardingSoapClient.java

```diff
@@ -0,0 +1,120 @@
+  private final SoapServiceProvider serviceProvider;
+
+  ForwardingSoapClient(SoapService service, SoapServiceProvider serviceProvider) {
+    this.serviceProvider = serviceProvider;
+    this.clientsCache = CacheBuilder.<WebServiceDefinition, SoapClient>newBuilder()
+        .expireAfterAccess(1, MINUTES)
+        .removalListener(new ForwardingClientRemovalListener())
+        .build(new SoapClientCacheLoader(service));
+  }
+
+  /**
+   * Returns a {@link SoapClient} instance connected to the {@link WebServiceDefinition} of the specified {@code id}.
+   *
+   * @param id the id of the {@link WebServiceDefinition}.
+   * @return a {@link SoapClient} instance
+   * @throws MuleException in any error case.
+   */
+  public SoapClient getSoapClient(String id) throws MuleException {
+    List<WebServiceDefinition> webServiceDefinitions = serviceProvider.getWebServiceDefinitions();
+    WebServiceDefinition wsd = webServiceDefinitions.stream().filter(ws -> ws.getServiceId().equals(id)).findAny()
+        .orElseThrow(() -> new IllegalArgumentException("Could not find a soap client id [" + id + "]"));
+    try {
+      return clientsCache.get(wsd);
+    } catch (ExecutionException e) {
+      throw new MuleRuntimeException(createStaticMessage("Error while retrieving soap client id [" + id + "]"), e);
+    }
+  }
+
+  /**
+   * Disconnects all the {@link SoapClient} instances created by this manager.
+   */
+  public void disconnect() {
+    clientsCache.invalidateAll();
+  }
+
+  /**
+   * {@link CacheLoader} implementation to load lazily {@link SoapClient}s.
+   */
+  private class SoapClientCacheLoader extends CacheLoader<WebServiceDefinition, SoapClient> {
+
+    private final SoapService service;
+
+    private SoapClientCacheLoader(SoapService service) {
+      this.service = service;
+    }
+
+    @Override
+    public SoapClient load(WebServiceDefinition definition) throws Exception {
+      SoapClientFactory clientFactory = service.getClientFactory();
+      SoapClientConfigurationBuilder configurationBuilder = builder()
+          .withService(definition.getService())
+          .withPort(definition.getPort())
+          .withWsdlLocation(definition.getWsdlUrl().toString());
+      if (definition.getAddress() != null) {
+        configurationBuilder.withAddress(definition.getAddress().toString());
+      }
+      serviceProvider.getSecurities().forEach(configurationBuilder::withSecurity);
+      return clientFactory.create(configurationBuilder.build());
+    }
+  }
+
+  /**
+   * {@link RemovalListener} implementation stop {@link SoapClient}s.
+   */
+  private class ForwardingClientRemovalListener implements RemovalListener<WebServiceDefinition, SoapClient> {
+
+    @Override
+    public void onRemoval(RemovalNotification<WebServiceDefinition, SoapClient> notification) {
+      try {
+        SoapClient client = notification.getValue();
+        if (client != null) {
+          client.stop();
+        }
+      } catch (MuleException e) {
```

Comment:
MuleException or Exception? just to avoid some cxf error to be propagated without the mule wrapper exception